### PR TITLE
Added support for french Bépo layout to Crkbd keeb

### PR DESCRIPTION
**Pull request by shinze**
Added support for french layout Bépo http://bepo.fr/

### CODE REVIEW

File: keyboards/crkbd/keymaps/shinze/keymap.c [hunk 1]
```diff
@@ -0,0 +1,269 @@
+#include QMK_KEYBOARD_H
+#include "bootloader.h"
```
Comment (noroadsleft) [inline keyboards/crkbd/keymaps/shinze/keymap.c]: `QMK_KEYBOARD_H` includes this file already.

```suggestion
```

File: keyboards/crkbd/keymaps/shinze/rules.mk [hunk 2]
```diff
@@ -0,0 +1,31 @@
+
+# Build Options
+#   change to "no" to disable the options, or define them in the Makefile in
+#   the appropriate keymap folder that will get included automatically
+#
+BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration(+1000)
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/shinze/rules.mk]: These can be condensed down into just:

```make
RGBLIGHT_ENABLE = yes
```

as the rest of them are unchanged from the keyboard-level `rules.mk`.

File: keyboards/crkbd/keymaps/shinze/rules.mk [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+NKRO_ENABLE = no            # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no      # Enable keyboard backlight functionality
+MIDI_ENABLE = no            # MIDI controls
+AUDIO_ENABLE = no           # Audio output on port C6
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = yes        # Enable WS2812 RGB underlight.  Do not enable this with audio at the same time.
+SWAP_HANDS_ENABLE = no        # Enable one-hand typing
+
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/shinze/rules.mk]:
```suggestion
```

File: keyboards/crkbd/keymaps/shinze/rules.mk [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/shinze/rules.mk]:
```suggestion
```

File: keyboards/crkbd/keymaps/shinze/rules.mk [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/shinze/rules.mk]:
```suggestion
```

File: keyboards/crkbd/keymaps/shinze/keymap.c [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+#define _ADJUST 16
+
+enum custom_keycodes {
+  BEPO = SAFE_RANGE,
+  LOWER,
+  RAISE,
+  ADJUST
+};
+
+enum macro_keycodes {
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/shinze/keymap.c]: Please remove this block - it's part of the old macro system which is deprecated, but you aren't using it anyway.

File: keyboards/crkbd/keymaps/shinze/keymap.c [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+extern rgblight_config_t rgblight_config;
+#endif
+
+extern uint8_t is_master;
+
+// Each layer gets a name for readability, which is then used in the keymap matrix below.
+// The underscores don't mean anything - you can have a layer called STUFF or any other name.
+// Layer names don't all need to be of the same length, obviously, and you can also skip them
+// entirely and just use numbers.
+#define _BEPO  0
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/shinze/keymap.c]: Non-consecutive layer numbers will generate extra empty layers in between, unnecessarily increasing flash usage. The better way to define layers is with an enum, eliminating the need to number them manually:

```c
enum layer_names {
    _BEPO,
    _LOWER,
    _RAISE,
    _ADJUST
};
```

File: keyboards/crkbd/keymaps/shinze/keymap.c [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+  LOWER,
+  RAISE,
+  ADJUST
+};
+
+enum macro_keycodes {
+  KC_SAMPLEMACRO,
+};
+
+#define _______ KC_TRNS
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/shinze/keymap.c]:
```suggestion
```

File: keyboards/crkbd/keymaps/shinze/keymap.c [hunk 9]
```diff
@@ -0,0 +1,269 @@
+#include QMK_KEYBOARD_H
+#include "bootloader.h"
+#include "keymap_french_osx.h"
+#include "keymap_bepo.h"
+#include "sendstring_bepo.h"
+#ifdef PROTOCOL_LUFA
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/shinze/keymap.c]: This block is probably not necessary.

File: keyboards/crkbd/keymaps/shinze/keymap.c (hunk 7 above)
Comment (noroadsleft) [inline keyboards/crkbd/keymaps/shinze/keymap.c]: Lines 27-30 should be replaced with:

```c
enum layer_names {
    _BEPO,
    _LOWER,
    _RAISE,
    _ADJUST,
};
```

Non-consecutive layer indexes will increase your firmware size for no benefit.

File: keyboards/crkbd/keymaps/shinze/keymap.c (hunk 8 above)
Comment (noroadsleft) [inline keyboards/crkbd/keymaps/shinze/keymap.c]: QMK defines this keycode alias automatically.

https://github.com/qmk/qmk_firmware/blob/a120a000d481ed14a87988383be46c5a76604bb6/quantum/quantum_keycodes.h#L29-L31

```suggestion
```

File: keyboards/crkbd/keymaps/shinze/config.h [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+//#define USE_MATRIX_I2C
+
+/* Select hand configuration */
+
+#define MASTER_LEFT
+// #define MASTER_RIGHT
+// #define EE_HANDS
+
+#define SSD1306OLED
```
Comment (drashna) [inline keyboards/crkbd/keymaps/shinze/config.h]: While this is fine, you may want to look into the `OLED_DRIVER_ENABLE` option.  It offers some more control over the OLED display, such as rotating the text. 

Take a look at my layout for an example of that. 

And you can see more info about the feature here:
https://docs.qmk.fm/#/feature_oled_driver

File: keyboards/crkbd/keymaps/shinze/keymap.c [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+    _______, _______, _______, _______, _______, _______,    _______, _______ ,_______ , _______, _______ , _______,
+  //,------+--------+--------+--------+--------+--------+   ,-------+--------+---------+--------+--------+---------.
+                               _______, _______, _______,    _______, _______, _______
+  )
+
+};
+
+int RGB_current_mode;
+
+void persistent_default_layer_set(uint16_t default_layer) {
```
Comment (drashna) [inline keyboards/crkbd/keymaps/shinze/keymap.c]: This function is depreciated, and has a core function that supersedes it.  

https://docs.qmk.fm/#/ref_functions?id=setting-the-persistent-default-layer

**Comment by drashna** [comment]
Any update on this?

**Comment by drashna** [comment]
Since it's been a long while without any updates or interaction here, we're closing this pull request. 

If you wish to update this, feel free to re-open the PR or resubmit it. 

If you need any help with this, don't hesitate to ask.

**Pull request closed without merging**